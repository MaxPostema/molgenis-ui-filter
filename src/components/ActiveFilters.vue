<template>
  <div>
    <button
      v-for="(item, key) in activeValues"
      :key="key"
      type="button"
      class="btn btn-light m-1 btn-outline-secondary"
      @click="removeFilter(item)"
    >
      {{ item.label }}: {{ item.value }}
      <font-awesome-icon
        icon="times"
        class="ml-1"
      />
    </button>
  </div>
</template>

<script>
import Vue from 'vue'

import { library } from '@fortawesome/fontawesome-svg-core'
import { faTimes } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faTimes)

export default Vue.extend({
  name: 'ActiveFilters',
  components: { FontAwesomeIcon },
  props: {
    filters: {
      type: Array,
      required: true
    },
    value: {
      type: Object,
      default: () => ({})
    }
  },
  data () {
    return {
      isOpen: this.collapsable ? !this.collapsed : true
    }
  },
  computed: {
    activeValues () {
      let values = []
      Object.keys(this.value).map(key => {
        const current = this.value[key]
        const filter = this.selectFilter(key)

        // Clean op the values by removing undefined entry's
        if (current === undefined) return

        // promise function has not settled yet
        if (typeof filter.options === 'function') return

        // Unpack array
        if (Array.isArray(current)) {
          // Checkbox
          if (filter.type === 'checkbox-filter') {
            current.map(subKey => {
              const option = filter.options.filter(item => item.value === subKey)[0]
              values.push({ key, subKey, value: option.text, label: filter.label })
            })
          }

          // Range Filter
          if (filter.type === 'range-filter') {
            values.push({ key, value: `${current[0]} to ${current[1]}`, label: filter.label })
          }
        } else {
          // Single item
          values.push({ key, value: current, label: filter.label })
        }
      })
      return values
    },
    splitArray () {
      return this.activeValues
    }
  },
  methods: {
    selectFilter (key) {
      return this.filters.filter(filter => filter.name === key)[0]
    },
    removeFilter ({ key, subKey }) {
      if (subKey === undefined) {
        let selections = { ...this.value }
        delete selections[key]
        this.$emit('input', selections)
      } else {
        let selections = { ...this.value }
        selections[key] = selections[key].filter(key => key !== subKey)
        this.$emit('input', selections)
      }
    }
  }
})
</script>

<style scoped>
  button svg path {
    transition: fill 0.3s;
  }
  button:hover svg path {
    fill: var(--danger);
  }
</style>
