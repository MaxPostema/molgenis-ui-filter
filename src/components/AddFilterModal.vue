<template>
  <div>
    <b-dropdown variant="primary" boundary="viewport" menu-class="shadow ml-2" ref="addFilter" dropright no-caret block>
      <template v-slot:button-content>
        Set filters <font-awesome-icon icon="caret-right" class="ml-2" />
      </template>
      <b-dropdown-text>
        Set filters
        <span class="float-right remove-button"
              @click.stop="$refs.addFilter.hide(true)"
        ><font-awesome-icon icon="times" /></span>
      </b-dropdown-text>
      <b-dropdown-form>
        <b-form-checkbox-group stacked v-model="selected" @input="changeFilters">
          <b-form-checkbox v-for="option in options" :key="option.name" :value="option.name" :class="{'ml-4':option.compound}">
            <span v-if="option.label">
              <span class="text-nowrap">{{option.label}} </span>
              <span class="text-secondary">
                <small><span class="text-nowrap">( {{option.name}} )</span></small>
              </span>
            </span>
            <span v-else>
              <span class="text-nowrap">{{option.name}} </span>
            </span>
            <span class="text-secondary">
              <small v-if="option.description"> - {{option.description}}</small>
            </span>
          </b-form-checkbox>
        </b-form-checkbox-group>
      </b-dropdown-form>
    </b-dropdown>
  </div>
</template>

<script>
import Vue from 'vue'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faCaretRight, faTimes } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faCaretRight, faTimes)

export default Vue.extend({
  name: 'AddFilterModal',
  components: { FontAwesomeIcon },
  props: {
    filters: {
      type: Array,
      required: true
    },
    value: {
      type: Array,
      default: () => []
    }
  },
  data () {
    return {
      selected: this.value
    }
  },
  computed: {
    options () {
      return this.filters.map(it => ({
        name: it.name,
        label: it.label,
        description: it.description,
        compound: it.compound
      }))
    }
  },
  methods: {
    remove () {

    },
    changeFilters () {
      this.$emit('input', this.selected)
    }
  }
})
</script>

<style scoped>
  .remove-button{
    transition: color 0.2s;
    height: inherit;
    width: 1.5em;
    text-align: center;
    display: inline-block;
    position: absolute;
    right: 10px;
    cursor: pointer;
  }
  .remove-button:hover{
    color: var(--danger);
  }
</style>
