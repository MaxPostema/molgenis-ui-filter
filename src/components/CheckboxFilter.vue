<template>
  <div>
    <fieldset :name="name" role="group" class="pt-2 b-form-group form-group">
      <div role="group" tabindex="-1" class="custom-controls-stacked">
        <div class="custom-control custom-checkbox" 
          v-for="(option, index) in visibleOptions" 
          :key="option.value"
        >
          <input :id="`option-${name}-${index}`"
            type="checkbox"
            :name="name"
            autocomplete="off"
            class="custom-control-input"
            :checked="value.includes(option.value)"
            @change="(event) => itemClick(option.value, event.target.checked)">
          <label :for="`option-${name}-${index}`" class="custom-control-label"><span>{{option.label}}</span></label>
        </div>
      </div>
    </fieldset>
    <a href="#" target="_self" class="toggle-slice card-link" @click.prevent="toggleSlice" v-if="showToggleSlice">{{toggleSliceText}}</a>
    <a href="#" target="_self" class="toggle-select card-link" @click.prevent="toggleSelect" >{{toggleSelectText}}</a>
  </div>
</template>

<style>
  .card-link { font-style: italic; font-size: small; }
</style>

<script>
export default {
  data () {
    return {
      collapsed: this.initiallyCollapsed,
      sliceOptions: this.maxVisibleOptions && this.options && this.maxVisibleOptions < this.options.length
    }
  },
  props: {
    name: String,
    label: String,
    options: Array,
    value: {
      type: Array,
      default: () => []
    },
    initiallyCollapsed: Boolean,
    maxVisibleOptions: Number
  },
  computed: {
    visibleOptions () {
      return this.sliceOptions ? this.options.slice(0, this.maxVisibleOptions) : this.options
    },
    showToggleSlice () {
      return this.maxVisibleOptions && this.maxVisibleOptions < this.options.length
    },
    toggleSelectText () {
      return this.value.length ? 'Deselect all' : 'Select all'
    },
    toggleSliceText () {
      return this.sliceOptions ? `Show ${this.options.length - this.maxVisibleOptions} more` : 'Show less'
    }
  },
  methods: {
    itemClick (value, checked) {
      if (checked) {
        this.$emit('input', [...this.value, value])
      } else {
        let leftover = this.value.filter(it => it !== value)
        if (leftover.length === 0) {
          leftover = undefined
        }
        this.$emit('input', leftover)
      }
    },
    toggleSelect () {
      this.$emit('input', this.value && this.value.length ? undefined : this.options.map(option => option.value))
    },
    toggleSlice () {
      this.sliceOptions = !this.sliceOptions
    }
  },
  watch: {
    options () {
      this.sliceOptions = this.showToggleSlice
    },
    maxVisibleOptions () {
      this.sliceOptions = this.showToggleSlice
    }
  }
}
</script>
