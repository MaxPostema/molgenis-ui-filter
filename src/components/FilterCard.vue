<template>
  <b-card
    no-body
    class="filter-card mb-2"
  >
    <b-card-header @click="collapsed = !collapsed">
      <font-awesome-icon
        icon="caret-right"
        :style="iconStyle"
        class="mr-2"
      />
      {{ label }}
    </b-card-header>
    <b-collapse
      :id="name"
      v-model="collapsed"
    >
      <b-card-body>
        <b-form-group :description="description">
          <slot />
        </b-form-group>
      </b-card-body>
    </b-collapse>
  </b-card>
</template>

<script>
export default {
  props: {
    name: {
      type: String,
      required: true
    },
    label: {
      type: String,
      required: true
    },
    description: {
      type: String,
      default: () => undefined
    }
  },
  data () {
    return {
      collapsed: this.initiallyCollapsed
    }
  },
  computed: {
    iconStyle () {
      return {
        transform: `rotate(${this.collapsed ? 90 : 0}deg)`,
        transition: 'transform 0.2s'
      }
    }
  }
}
</script>

<style scoped>
  .card-header {
    cursor: pointer;
  }
</style>
