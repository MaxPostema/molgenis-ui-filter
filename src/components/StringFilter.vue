<template>
  <b-input-group>
    <b-form-input
      :name="name"
      v-model="model"
      :placeholder="placeholder"
      trim
    >
    </b-form-input>
    <b-input-group-append>
      <b-button variant="outline-secondary" @click.prevent="model=''">x</b-button>
    </b-input-group-append>
  </b-input-group>
</template>

<script>
export default {
  name: 'StringFilter',
  props: {
    name: String,
    placeholder: String,
    value: String
  },
  computed: {
    model: {
      get () {
        return this.value || ''
      },
      set (value) {
        this.$emit('input', value === '' ? undefined : value)
      }
    }
  }
}
</script>
