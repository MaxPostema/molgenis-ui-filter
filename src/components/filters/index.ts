import BootstrapVue from 'bootstrap-vue'
import Vue from 'vue'
Vue.use(BootstrapVue)

export { default as CheckboxFilter } from './CheckboxFilter.vue'
export { default as StringFilter } from './StringFilter.vue'
