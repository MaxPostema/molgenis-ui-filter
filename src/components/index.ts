import BootstrapVue from 'bootstrap-vue'
import Vue from 'vue'
Vue.use(BootstrapVue)

export { default as FilterCard } from './FilterCard.vue'
export { default as FilterContainer } from './FilterContainer.vue'
export { default as ChangeFilters } from './ChangeFilters.vue'
export { default as ActiveFilters } from './ActiveFilters.vue'
