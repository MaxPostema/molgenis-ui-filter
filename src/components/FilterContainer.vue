<template>
  <div
    class="filter-container"
    @mouseup="drag=false"
  >
    <draggable
      v-model="filtersToShow"
      :disabled="!canEdit"
      :class="canEdit?(drag?'dragdrop dragging':'dragdrop'):''"
      @choose="drag=true"
      @end="drag=false"
      @input="selectionUpdate"
    >
      <transition-group>
        <filter-card
          v-for="filter in listOfVisibleFilters"
          :key="filter.name"
          v-bind="filter"
          :can-remove="canEdit"
          @removeFilter="removeFilter"
        >
          <component
            :is="filter.type"
            :value="value[filter.name]"
            v-bind="filter"
            @input="value => selectionChange(filter.name, value)"
          />
        </filter-card>
      </transition-group>
    </draggable>

    <add-filter-modal
      v-if="listOfInvisibleFilters.length>0"
      v-model="filtersToShow"
      :filters="listOfInvisibleFilters"
      @input="selectionUpdate"
    />
  </div>
</template>

<script>
import AddFilterModal from './AddFilterModal.vue'
import { FilterCard } from '.'
import { StringFilter, CheckboxFilter } from './filters/'
import draggable from 'vuedraggable'

export default {
  name: 'FilterContainer',
  components: { AddFilterModal, StringFilter, CheckboxFilter, FilterCard, draggable },
  props: {
    filters: {
      type: Array,
      required: true
    },
    value: {
      type: Object,
      default: () => ({})
    },
    filtersShown: {
      type: Array,
      required: false,
      default: () => []
    },
    canEdit: {
      type: Boolean,
      required: false,
      default: () => false
    }
  },
  data () {
    return {
      filtersToShow: this.filtersShown,
      filterToAdd: null,
      drag: false
    }
  },
  computed: {
    listOfVisibleFilters () {
      return this.filtersToShow.map(id => this.filters.find(filter => filter.name === id))
    },
    listOfInvisibleFilters () {
      return this.filters.filter(filter => !this.filtersToShow.includes(filter.name))
    }
  },
  methods: {
    removeFilter (name) {
      this.filtersToShow = this.filtersToShow.filter(filter => name !== filter)
      this.$emit('update', this.filtersToShow)

      let selections = { ...this.value }
      delete selections[name]
      this.$emit('input', selections)
    },
    selectionChange (name, value) {
      this.$emit('input', { ...this.value, [name]: value })
    },
    selectionUpdate () {
      this.$emit('update', this.filtersToShow)
    }
  }
}
</script>

<style scoped>
  .dragdrop{
    cursor: grab;
  }
  .dragdrop.dragging{
    cursor: grabbing;
  }
</style>
