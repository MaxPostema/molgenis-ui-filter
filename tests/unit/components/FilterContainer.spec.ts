import { mount } from '@vue/test-utils'
import { FilterContainer } from '@/components'

describe('FilterContainer.vue', () => {
  const optionsPromise = () => {
    return new Promise(
      function (resolve) {
        resolve([{ value: 'red', text: 'Red' }, { value: 'green', text: 'Green' }, { value: 'blue', text: 'Blue' }])
      }
    )
  }
  const options = {
    stubs: {
      'font-awesome-icon': '<div />'
    },
    propsData: {
      value: {
        string: 'blah',
        checkbox: ['red']
      },
      filters: [ {
        name: 'string',
        label: 'String',
        description: 'search by string',
        placeholder: 'placeholder',
        type: 'string-filter',
        collapsable: false
      }, {
        name: 'checkbox',
        label: 'Checkbox',
        collapsed: false,
        bulkOperation: true,
        options: optionsPromise,
        type: 'checkbox-filter'
      }, {
        name: 'checkbox2',
        label: 'Checkbox',
        collapsed: false,
        bulkOperation: true,
        options: optionsPromise,
        type: 'checkbox-filter'
      }],
      filtersShown: ['string', 'checkbox']
    }
  }

  let wrapper = mount(FilterContainer, options)

  beforeEach(() => {
    wrapper = mount(FilterContainer, options)
  })

  it('matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('consolidates all filter output and sends them via input event', () => {
    wrapper.find('input[name="string"]').setValue('test')
    wrapper.find('input[value="red"]').trigger('click')
    expect(wrapper.emitted().input).toEqual([
      [{ 'checkbox': ['red'], 'string': 'test' }],
      [{ 'checkbox': undefined, 'string': 'blah' }]
    ])
  })

  it('cannot remove filters by default', () => {
    expect(wrapper.find('.remove-button').exists()).toBe(false)
  })

  it('can remove filters when editable', () => {
    wrapper.setProps({ canEdit: true })
    wrapper.find('.remove-button').trigger('click')
    expect(wrapper.emitted().update).toEqual([[['checkbox']]])
    expect(wrapper.emitted().input).toEqual([[{ checkbox: [ 'red' ] }]])
  })

  it('cannot add filters by default', () => {
    expect(wrapper.find('.add-button').exists()).toBe(false)
  })

  it('can add filters when editable', (done) => {
    wrapper.setProps({ canEdit: true })
    wrapper.find('.add-button').trigger('click')
    wrapper.vm.$nextTick(() => {
      wrapper.find('#modal-add-filter select[name=filter]').setValue('checkbox2')
      wrapper.find('#modal-add-filter .modal-footer button.btn-primary').trigger('click')
      expect(wrapper.emitted().update[0]).toEqual([ [ 'string', 'checkbox', 'checkbox2' ] ])
      done()
    })
  })

  it('shows the filters in the correct order ( based on filtersToShow ) ', () => {
    expect((wrapper.vm as any).listOfVisibleFilters[0].label).toBe('String')
    expect((wrapper.vm as any).listOfVisibleFilters[1].label).toBe('Checkbox')
    wrapper.setData({ filtersToShow: ['checkbox', 'string'] }) // flip order
    expect((wrapper.vm as any).listOfVisibleFilters[0].label).toBe('Checkbox')
    expect((wrapper.vm as any).listOfVisibleFilters[1].label).toBe('String')
  })
})
