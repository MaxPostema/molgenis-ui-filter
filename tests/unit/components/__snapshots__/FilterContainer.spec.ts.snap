// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterContainer.vue matches the snapshot 1`] = `
<div
  class="filter-container"
>
  <div
    class=""
  >
    <span>
      <div
        class="card filter-card mb-2"
        placeholder="placeholder"
        type="string-filter"
      >
        <!---->
        <!---->
        <div
          class="card-header"
        >
          <div
            class="title mr-3 px-1"
            title="String"
          >
            <!---->
            
      String
      
            <!---->
          </div>
        </div>
         
        <div
          class="collapse show"
          id="string"
        >
          <div
            class="card-body"
          >
            <!---->
            <!---->
            <fieldset
              aria-describedby="__BVID__30__BV_description_"
              class="form-group"
              id="__BVID__30"
            >
              <!---->
              <div
                role="group"
                tabindex="-1"
              >
                <div
                  class="input-group"
                  description="search by string"
                  label="String"
                  role="group"
                  type="string-filter"
                >
                  <!---->
                  <!---->
                  <input
                    class="form-control"
                    id="__BVID__32"
                    name="string"
                    placeholder="placeholder"
                    type="text"
                  />
                   
                  <div
                    class="input-group-append"
                  >
                    <button
                      class="btn btn-outline-secondary"
                      type="button"
                    >
                      <div
                        icon="times"
                      />
                    </button>
                  </div>
                  <!---->
                  <!---->
                </div>
                <!---->
                <!---->
                <small
                  class="form-text text-muted"
                  id="__BVID__30__BV_description_"
                  tabindex="-1"
                >
                  search by string
                </small>
              </div>
            </fieldset>
          </div>
        </div>
        <!---->
        <!---->
      </div>
      <div
        bulkoperation="true"
        class="card filter-card mb-2"
        options="[object Object],[object Object],[object Object]"
        type="checkbox-filter"
      >
        <!---->
        <!---->
        <div
          class="card-header collapsable"
        >
          <div
            class="title mr-3 px-1"
            title="Checkbox"
          >
            <div
              class="mr-2"
              icon="caret-right"
              style="transform: rotate(90deg); transition: transform 0.2s;"
            />
            
      Checkbox
      
            <!---->
          </div>
        </div>
         
        <div
          class="collapse show"
          id="checkbox"
        >
          <div
            class="card-body"
          >
            <!---->
            <!---->
            <fieldset
              class="form-group"
              id="__BVID__38"
            >
              <!---->
              <div
                role="group"
                tabindex="-1"
              >
                <div
                  type="checkbox-filter"
                >
                  <div
                    class=""
                    id="__BVID__40"
                    role="group"
                    tabindex="-1"
                  >
                    <div
                      class="custom-control custom-checkbox"
                    >
                      <input
                        autocomplete="off"
                        class="custom-control-input"
                        id="__BVID__40__BV_option_0_"
                        name="checkbox"
                        type="checkbox"
                        value="red"
                      />
                      <label
                        class="custom-control-label"
                        for="__BVID__40__BV_option_0_"
                      >
                        <span>
                          Red
                        </span>
                      </label>
                    </div>
                    <div
                      class="custom-control custom-checkbox"
                    >
                      <input
                        autocomplete="off"
                        class="custom-control-input"
                        id="__BVID__40__BV_option_1_"
                        name="checkbox"
                        type="checkbox"
                        value="green"
                      />
                      <label
                        class="custom-control-label"
                        for="__BVID__40__BV_option_1_"
                      >
                        <span>
                          Green
                        </span>
                      </label>
                    </div>
                    <div
                      class="custom-control custom-checkbox"
                    >
                      <input
                        autocomplete="off"
                        class="custom-control-input"
                        id="__BVID__40__BV_option_2_"
                        name="checkbox"
                        type="checkbox"
                        value="blue"
                      />
                      <label
                        class="custom-control-label"
                        for="__BVID__40__BV_option_2_"
                      >
                        <span>
                          Blue
                        </span>
                      </label>
                    </div>
                  </div>
                   
                  <span>
                    <!---->
                     
                    <a
                      class="toggle-select card-link"
                      href="#"
                      target="_self"
                    >
                      
      Deselect all
    
                    </a>
                  </span>
                </div>
                <!---->
                <!---->
                <!---->
              </div>
            </fieldset>
          </div>
        </div>
        <!---->
        <!---->
      </div>
    </span>
  </div>
   
  <!---->
</div>
`;
